src/main/java/com/google/cloud/tools/maven/DeployAllMojo.java

```diff
@@ -38,17 +38,51 @@ public void execute() throws MojoExecutionException, MojoFailureException {
       deployables.clear();
     }
 
-    String[] validYamls = {
-      "app.yaml", "cron.yaml", "dispatch.yaml", "dos.yaml", "index.yaml", "queue.yaml"
-    };
-    for (String yamlName : validYamls) {
-      File yaml = stagingDirectory.toPath().resolve(yamlName).toFile();
-      if (yaml.exists()) {
-        getLog().info("deployAll: Preparing to deploy " + yamlName);
-        deployables.add(yaml);
+    String[] configYamls = {"cron.yaml", "dispatch.yaml", "dos.yaml", "index.yaml", "queue.yaml"};
+    File appYaml = stagingDirectory.toPath().resolve("app.yaml").toFile();
+    if (appYaml.exists()) {
+      addDeployable(appYaml);
+    }
+
+    if (isStandardStaging()) {
+      // Check staging directory for app.yaml
+      if (!appYaml.exists()) {
+        throw new MojoExecutionException("Failed to deploy all: app.yaml not found.");
+      }
+
+      // Check staging directory for config yamls
+      for (String yamlName : configYamls) {
+        File yaml = stagingDirectory.toPath().resolve(yamlName).toFile();
+        if (yaml.exists()) {
+          addDeployable(yaml);
+        }
+      }
+    } else {
+      if (!appYaml.exists()) {
+        // Check source directory if app.yaml is missing
+        appYaml = appEngineDirectory.toPath().resolve("app.yaml").toFile();
+        if (appYaml.exists()) {
+          addDeployable(appYaml);
+        } else {
+          throw new MojoExecutionException("Failed to deploy all: app.yaml not found.");
+        }
+      }
+
+      // Look appengine directory for config yamls
+      configureAppEngineDirectory();
+      for (String yamlName : configYamls) {
```

Comment:
nit: This could probably also be combined with the loop for standard, just changing the directory it resolves for.